middleware/plugin/ngx_http_casual_module.c

```diff
@@ -0,0 +1,688 @@
+   }
+   else
+   {
+      // same content-type as incomming
+      r->headers_out.content_type.data = ngx_pstrdup( r->pool, &context->content_type);
+      r->headers_out.content_type.len = context->content_type.len;
+   }
+
+   set_custom_header_in_headers_out( r, &reply.headers);
+
+   if( ngx_http_send_header(r) != NGX_OK)
+   {
+      ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "casual: ngx_http_send_header(r) != NGX_OK");
+      return NGX_HTTP_INTERNAL_SERVER_ERROR;
+   }
+
+   if( r->headers_out.content_type.len != 0)
+   {
+      if( ngx_http_output_filter(r, response) != NGX_OK)
+      {
+         ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "casual: ngx_http_output_filter() != NGX_OK");
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+   }
+
+   return r->headers_out.status;
+}
+
+static void request_cleanup( void* data)
+{
+   ngx_http_casual_ctx_t* context = data;
+   ngx_log_error( NGX_LOG_NOTICE, context->r->connection->log, 0, "casual: %s", __FUNCTION__);
+ 
+   if( context->rev.active)
+   {
+      // context->rev.handler = empty_handler;
+      if (NGX_OK != ngx_del_event(&context->rev, NGX_READ_EVENT, NGX_CLOSE_EVENT))
+      {
+         ngx_log_error(NGX_LOG_ERR, context->r->connection->log, 0, "casual: %s, ngx_del_event(rev, NGX_READ_EVENT, NGX_CLOSE_EVENT)", __FUNCTION__);
+      }
+   }
+   if( context->wev.active)
+   {
+      // context->wev.handler = empty_handler;
+      if (NGX_OK != ngx_del_event(&context->wev, NGX_WRITE_EVENT, NGX_CLOSE_EVENT))
+      {
+         ngx_log_error(NGX_LOG_ERR, context->r->connection->log, 0, "casual: %s, ngx_del_event(wev, NGX_WRITE_EVENT, NGX_CLOSE_EVENT)", __FUNCTION__);
+      }
+   }
+   if( context->timer.timer_set)
+   {
+      ngx_del_timer(&context->timer);
+   }
+
+   casual_http_inbound_deallocate( &context->casual_handle);
+}
+
+static void initilize_context( ngx_http_casual_ctx_t* context)
+{
+   context->casual_handle.fd = 0;
+   context->casual_handle.context_holder = NULL;
+}
+
+static ngx_int_t get_header_length( ngx_http_request_t* r)
+{
+   ngx_list_part_t *part = &r->headers_in.headers.part;
+
+   ngx_uint_t length = 0;
+   ngx_uint_t i = 0;
+
+   for( i = 0; /* void */; i++)
+   {
+      if( i >= part->nelts)
+      {
+         if( part->next == NULL)
+         {
+            /* The last part, search is done. */
+            break;
+         }
+
+         part = part->next;
+         i = 0;
+      }
+      length++;
+   }
+   return length;
+}
+
+static void copy_headers( ngx_http_request_t* r, casual_http_inbound_header_t* header)
+{
+   ngx_list_part_t *part;
+   ngx_table_elt_t *h;
+   ngx_uint_t i;
+
+   /*
+    Get the first part of the list. There is usual only one part.
+    */
+   part = &r->headers_in.headers.part;
+   h = part->elts;
+
+   /*
+    Headers list array may consist of more than one part,
+    so loop through all of it
+    */
+   casual_http_inbound_header_t* headerpointer = header;
```

Comment:
A assume this is done just to "not effect caller _header_"? It will not be effected, since `header` is passad by value.